### PR TITLE
remove compile mode from trainer and add compatibility for callbacks

### PR DESCRIPTION
**Pull request by nw13slx**
## Description

- always save state_dict in trainer even if the model is compiled.
- allow skipping gradient calculations
- move load_callable to utils
- store the number of types in per_species_shifts
- 

## Motivation and Context

Resolves: #???

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [ ] My code follows the code style of this project and has been formatted using `black`.
- [ ] All new and existing tests passed, including on GPU (if relevant).
- [ ] I have added tests that cover my changes (if relevant).
- [ ] The option documentation (`docs/options`) has been updated with new or changed options.
- [ ] I have updated `CHANGELOG.md`.
- [ ] I have updated the documentation (if relevant).

### CODE REVIEW

File: nequip/train/trainer.py [hunk 1]
```diff
@@ -883,13 +884,13 @@ def epoch_step(self):
                     )
                     self.end_of_batch_log(batch_type=category)
                     for callback in self.end_of_batch_callbacks:
-                        callback(self)
+                        load_callable(callback)(self)
```
Comment (Linux-cpp-lisp) [inline nequip/train/trainer.py:887]: can't we `self.end_of_batch_callbacks = [load_callable(c) for c in end_of_batch_callbacks]` in `__init__` instead?
seems wasteful to call down into YAML every single epoch

similar point for the other callback lists

Comment (nw13slx) [reply nequip/train/trainer.py:887]: yes. I need to rename the variable though. otherwise the dict saving will save the functions instead of the arguments.

File: nequip/nn/_atomwise.py [hunk 2]
```diff
@@ -118,6 +118,7 @@ def __init__(
         irreps_in={},
     ):
         super().__init__()
+        self.num_types = num_types
```
[the remark this answers is not in the document]
Comment (nw13slx) [reply nequip/nn/_atomwise.py:121]: this is necessary if we want to do some callback functions that need to know about the number of types. Is there other route to retrieve this value from trainer?

[the remark this answers is not in the document]
Comment (nw13slx) [reply nequip/nn/_atomwise.py:121]: it is fixed as suggested

File: nequip/train/trainer.py (hunk 1 above)
Comment (Linux-cpp-lisp) [reply nequip/train/trainer.py]: aha
one easier option maybe is just to memoize the load_callback function?
have it keep a dict so it doesn't call to load YAML repeatedly @nw13slx

File: nequip/nn/_atomwise.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Linux-cpp-lisp) [reply nequip/nn/_atomwise.py:121]: :+1: this is just fine was only wondering

File: nequip/train/trainer.py (hunk 1 above)
Comment (nw13slx) [reply nequip/train/trainer.py]: sorry. my previous message was prefixed. I have revised it as you suggested.

Comment (nw13slx) [reply nequip/train/trainer.py]: just use something like "self._end_of_xxx_callbacks" to store the functions

Comment (Linux-cpp-lisp) [reply nequip/train/trainer.py]: great
